feat(android+ios): implement payment callbacks

mortend opened the pull request:
This makes it possible to pass custom payment callbacks from TypeScript.

This was possible before using version 0.5.0 of this module, but the
functionality no longer exists after replacing AdyenDropInView.

We depend on this functionality in our app, so this patch is an effort
to bring it back.

mortend committed:
feat(android+ios): implement payment callbacks

This makes it possible to pass custom payment callbacks from TypeScript.

This was possible before using version 0.5.0 of this module, but the
functionality no longer exists after replacing AdyenDropInView.

We depend on this functionality in our app, so this patch is an effort
to bring it back.

jesperjohansson commented on ios/AdyenDropInModule.swift:
```diff
@@ -112,6 +112,74 @@ class AdyenDropInModule: NSObject {
+    @objc
+    func setSubmitCallback(_ onSubmit: @escaping RCTResponseSenderBlock) {
+        print("Called setSubmitCallback")
+        self.onSubmitCallback = onSubmit
```
Looks like
`private var onSubmitCallback: RCTResponseSenderBlock? = nil`
and
`private var onAdditionalDetailsCallback: RCTResponseSenderBlock? = nil`
are missing from the PR

mortend replied:
Good catch! It looks like those parts fell off somewhere along the way... Updated now.

jesperjohansson reviewed (changes requested):
Hello! Thank you for contributing! 🎉

The reason we removed this functionality was because Android would sometimes pause the JS execution thread after the Adyen activity had been launched. This would result in bugs like the JS callbacks being unreachable and the payment would be stuck in the activity (spinner stuck when the JS callback should be executed). I couldn’t find a reliable workaround so I decided to switch to using native.

In our case we refactored the JS part in our app so the submit and additional details requests are handled by the native module and instead use the custom endpoints config and the custom headers config (auth token). We also changed from using a view to a button component to keep it React-like, like in the example app.

Are you sure you need the callback-approach?

mortend commented:
> Hello! Thank you for contributing! 🎉
> 
> The reason we removed this functionality was because Android would sometimes pause the JS execution thread after the Adyen activity had been launched. This would result in bugs like the JS callbacks being unreachable and the payment would be stuck in the activity (spinner stuck when the JS callback should be executed). I couldn’t find a reliable workaround so I decided to switch to using native.
> 
> In our case we refactored the JS part in our app so the submit and additional details requests are handled by the native module and instead use the custom endpoints config and the custom headers config (auth token). We also changed from using a view to a button component to keep it React-like, like in the example app.
> 
> Are you sure you need the callback-approach?

Hello! :)

OK, thank you for this information.

This implies that also our app is probably suffering from bugs like this on Android since we're using TypeScript callbacks.

As a first step towards upgrading to the latest version it seemed easier to bring back this functionality and keep our existing TypeScript callbacks as-is.

mortend commented:
We'll still need a callback in TypeScript in our app even after porting the submit and additional details requests. This is because we want to show a "please wait" screen behind the drop-in in RN-land. Currently we do that in `onSubmit`.

* If this callback sometimes doesn't fire on Android it is not a critical error. The app and payment will still work fine with all requests being done in native code, if I understand correctly.

Regarding pausing of the JavaScript thread on Android:

A) Does it happen with Credit Card payments inside the drop-in?
B) Does it happen when opening other apps from the drop-in?
C) Both (or all kinds of payments)?

jesperjohansson commented:
> We'll still need a callback in TypeScript in our app even after porting the submit and additional details requests. This is because we want to show a "please wait" screen behind the drop-in in RN-land. Currently we do that in `onSubmit`.
> 
> * If this callback sometimes doesn't fire on Android it is not a critical error. The app and payment will still work fine with all requests being done in native code, if I understand correctly.
> 
> Regarding pausing of the JavaScript thread on Android:
> 
> A) Does it happen with Credit Card payments inside the drop-in? B) Does it happen when opening other apps from the drop-in? C) Both (or all kinds of payments)?

I know it happens with credit card payments. I have had it happen in `onSubmit` before any page/app was opened. It probably happens with other payment methods too.
